### PR TITLE
CAMEL-11959: Add new camel-yql component

### PR DESCRIPTION
**Pull request by carlphilipp**
Add a new YQL component

### CODE REVIEW

**Comment by oscerd** [comment]
We will review soon, thanks @carlphilipp :-)

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlComponent.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.camel.component.yql.configuration.YqlConfigurationValidator;
+import org.apache.camel.impl.DefaultComponent;
+
+public class YqlComponent extends DefaultComponent {
+
+    @Override
+    protected Endpoint createEndpoint(final String uri, final String remaining, final Map<String, Object> parameters) throws Exception {
+        final YqlConfiguration configuration = new YqlConfiguration();
+        configuration.setQuery(remaining);
+        configuration.setFormat((String) parameters.getOrDefault("format", "json"));
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlComponent.java]: Camel can do this automatic in the setProperties(configuration, parameters) you do further below. If you want to set default values then set them on the configuration class instead

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+@UriParams
+public class YqlConfiguration {
+
+    @UriPath
+    @Metadata(required = "true")
+    private String query;
+
+    @UriParam
+    private String format;
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java]: Here you can set the default values.

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import org.apache.camel.Consumer;
+import org.apache.camel.Processor;
+import org.apache.camel.Producer;
+import org.apache.camel.component.yql.configuration.YqlConfiguration;
+import org.apache.camel.impl.DefaultEndpoint;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriParam;
+
+@UriEndpoint(firstVersion = "2.21.0", scheme = "yql", title = "Yahoo Query Language", syntax = "yql:query", producerOnly = true, label = "api,http")
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlEndpoint.java]: The labels should likely be something better to caregorize what this component is

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+public class YqlClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(YqlClient.class);
+    private static final CloseableHttpClient HTTP_CLIENT = HttpClients.createDefault();
+
+    public YqlResponse get(final String query, final String format, final boolean diagnostics, final String callback) throws Exception {
+
+        final URI uri = new URIBuilder()
+                .setScheme("http")
+                .setHost("query.yahooapis.com")
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java]: Would you ever need to configure the host and path? If so then we can make them configurable on the component level.

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                .setPath("/v1/public/yql")
+                .setParameter("format", format)
+                .setParameter("diagnostics", Boolean.toString(diagnostics))
+                .setParameter("env", "store://datatables.org/alltableswithkeys")
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java]: Should env not be configurable?

File: components/camel-yql/src/test/resources/log4j2.properties [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+## ---------------------------------------------------------------------------
+
+appender.file.type = File
+appender.file.name = file
+appender.file.fileName = target/camel-sql-component-test.log
```
Comment (davsclaus) [inline components/camel-yql/src/test/resources/log4j2.properties]: The file name should be the component name

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class YqlClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(YqlClient.class);
+    private static final CloseableHttpClient HTTP_CLIENT = HttpClients.createDefault();
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java]: Can we avoid a static shared client in the JVM? This dont work well on app servers etc. Is the HTTP_CLIENT thread safe also? If it should be shared/reused then we can have a http client on the component that they producers can reuse.

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    public void setDiagnostics(final boolean diagnostics) {
+        this.diagnostics = diagnostics;
+    }
+
+    public String getCallback() {
+        return callback;
+    }
+
+    /**
+     * Set the callback function
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java]: Can you add some more documentation what this callback function is used for? And also put some more into the adoc file about this, and maybe with an example.

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlProducer.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        final YqlConfiguration configuration = endpoint.getConfiguration();
+        final YqlResponse yqlResponse = yahooClient.get(
+                configuration.getQuery(),
+                configuration.getFormat(),
+                configuration.isDiagnostics(),
+                configuration.getCallback()
+        );
+        exchange.getIn().setHeader(CAMEL_YQL_HTTP_STATUS, yqlResponse.getStatus());
+        exchange.getIn().setHeader(CAMEL_YQL_HTTP_REQUEST, yqlResponse.getHttpRequest());
+        exchange.getIn().setBody(yqlResponse.getBody());
```
Comment (davsclaus) [inline components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlProducer.java]: I wonder if there was an error, eg status code is 5xx, 4xx etc if we should not throw an exception instead by default. And then allow to turn this on|off - this is how the http components does it. Then its easier for Camel users to know if something went wrong.

**Comment by davsclaus** [comment]
Great work, I added some comments which would be good if you could take a look at.

**Comment by provCarlHarmant** [comment]
Working on the comments

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlComponent.java (hunk 1 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlComponent.java]: Done

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlEndpoint.java (hunk 3 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlEndpoint.java]: is `label="yql"` ok?

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlProducer.java (hunk 9 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/YqlProducer.java]: Updated that part of the code with a flag to turn the feature on|off

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java (hunk 7 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java]: Yes `CloseableHttpClient` is thread safe. I added a connection manager that can be set on the component.

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java (hunk 5 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/client/YqlClient.java]: Yes it should, updated

File: components/camel-yql/src/test/resources/log4j2.properties (hunk 6 above)
Comment (carlphilipp) [reply components/camel-yql/src/test/resources/log4j2.properties]: Fixed

File: components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java (hunk 8 above)
Comment (carlphilipp) [reply components/camel-yql/src/main/java/org/apache/camel/component/yql/configuration/YqlConfiguration.java]: Added a bunch of comments

**Comment by davsclaus** [comment]
Let us know when you think this is ready

**Comment by carlphilipp** [comment]
It should be ready for review. Thanks!

**Comment by davsclaus** [comment]
Thanks I am working on merging this now

**Pull request closed without merging**